### ISSUE TITLE
Feature: override coursier download location

### ISSUE THREAD

**Issue by peter-janssen** [bazelbuild/rules_jvm_external#389]
Currently coursier is retrieved from: https://github.com/coursier/coursier/releases/download/{COURSIER_CLI_VERSION}/coursier.jar
This might not always be possible due to corporate proxies etc. 

The possibility to override this url to a private repository would be nice.

**Comment by illicitonion** [comment]
Does anyone (cc @jin) have an idea for what the interface for this should be?

**Comment by jin** [comment]
Since we're already using an environment variable for `COURSIER_CACHE`, `COURSIER_URL` and `COURSIER_SHA256` would work to replace the values here:

https://github.com/bazelbuild/rules_jvm_external/blob/bad9e2501279aea5268b1b8a5463ccc1be8ddf65/coursier.bzl#L571-L574

**Comment by wisechengyi** [comment]
Thanks for the pointers, @jin!

However I've tried the following but it still tried to reach the default urls.
```
./bazel build --repo_env='COURSIER_URL=https://<some internal host>/io/get-coursier/bazel/v2.0.0-RC5-3/coursier.jar'
```

Looking back at the bzl files:
```
$ rg COURSIER_CACHE
coursier.bzl
470:    location = env_dict.get("COURSIER_CACHE")
472:    # If we're not using the unsafe shared cache, or if COURSIER_CACHE is not defined,
548:        # If not using the shared cache and the user did not specify a COURSIER_CACHE, set the default
552:        environment = {"COURSIER_CACHE": str(repository_ctx.path(coursier_cache_location))}
938:        "COURSIER_CACHE",
```

`COURSIER_CACHE` is used, but didn't see `COURSIER_URL` getting picked up anywhere.

(sorry newbie to bazel here in case I missed any basics)

**Comment by jin** [comment]
This hasn't been implemented yet. I was responding to @illicitonion's question on how this would look like, design-wise.

**Comment by wisechengyi** [comment]
Ah thanks for clarifying! I will sync with @illicitonion and try to take a stab at it.

**Comment by illicitonion** [comment]
@jin Skimming through the code, how would you feel about `maven_install` taking an optional `coursier_deploy_jar` arg, which would default to `@rules_jvm_external//:{COURSIER_CLI_HTTP_FILE_NAME}` but could be overridden in the user's `WORKSPACE` file to some other `http_file` if desired? I guess the env var doesn't allow changing the sha, which means the rules can be sure the correct version is being used, but env vars generally feel more fiddly to me than targets...

@wisechengyi If we go the env var way, I think the two stages would be roughtly:
1. Add a new env var to the list of env vars the rule is allowed to consume here: https://github.com/bazelbuild/rules_jvm_external/blob/bad9e2501279aea5268b1b8a5463ccc1be8ddf65/coursier.bzl#L930
2. If the env var is set (which can be accessed via `respository_ctx.os.environ`), use it here https://github.com/bazelbuild/rules_jvm_external/blob/bad9e2501279aea5268b1b8a5463ccc1be8ddf65/coursier.bzl#L571 instead of the hard-coded list of two URLs.

**Comment by jin** [comment]
I'd prefer not to, because coursier is an implementation detail of the
`maven_install` API.

Generally, things that would result in a change of the generated dependency
graph should go into maven_install, and things that affect how the
machinery works are in env vars (e.g. coursier cache, http proxy, etc)

I don't expect most users to change the coursier deploy jar anyway, so env
vars should be good enough for now.

**Comment by illicitonion** [comment]
Makes sense, thanks!

**Comment by jin** [comment]
Fixed by @wisechengyi in #402. Thanks!
